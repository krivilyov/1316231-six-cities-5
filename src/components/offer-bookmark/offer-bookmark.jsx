import React from "react";
import {fetchBookmarkOffers, updateOfferBookmarkStatus} from "../../store/api-actions";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import {getIsAuthorizedStatus} from "../../store/selectors";
import {AppRoute, offerBookmarkTypeProperty} from "../../const";
import OfferBookmarkContent from "../offer-bookmark-content/offer-bookmark-content";

const OfferBookmark = (props) => {
  const {offerId, isAuthorizedStatus, offerBookmarkStatus, onChangeBookmark} = props;

  const handleBookmarkClick = (evt) => {
    evt.preventDefault();
    onChangeBookmark(offerId, !offerBookmarkStatus);
  };

  return isAuthorizedStatus ? (
    <button className={`property__bookmark-button ${
      offerBookmarkStatus ? ` property__bookmark-button--active ` : ``}button`}
    type="button"
    onClick={handleBookmarkClick}
    >
      <OfferBookmarkContent
        offerBookmarkStatus={offerBookmarkStatus}
        offerBookmarkTypeProperty={offerBookmarkTypeProperty.page}
      />
    </button>
  ) : (
    <Link to={AppRoute.LOGIN} className={`property__bookmark-button ${
      offerBookmarkStatus ? ` property__bookmark-button--active ` : ``}button`}
    type="button"
    >
      <OfferBookmarkContent
        offerBookmarkStatus={offerBookmarkStatus}
        offerBookmarkTypeProperty={offerBookmarkTypeProperty.page}
      />
    </Link>
  );
};

OfferBookmark.propTypes = {
  offerId: PropTypes.string.isRequired,
  isAuthorizedStatus: PropTypes.bool.isRequired,
  offerBookmarkStatus: PropTypes.bool.isRequired,
  onChangeBookmark: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  isAuthorizedStatus: getIsAuthorizedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onChangeBookmark(offerId, bookmarkStatus) {
    dispatch(updateOfferBookmarkStatus(offerId, bookmarkStatus));
    dispatch(fetchBookmarkOffers());
  },
});

export {OfferBookmark};
export default connect(mapStateToProps, mapDispatchToProps)(OfferBookmark);
