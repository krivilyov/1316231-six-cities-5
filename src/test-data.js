export const testInitialState = {
  COMMON: {
    activeCity: `Amsterdam`,
    offers: [],
    reviews: [],
    sortingType: `popular`,
    mouseOverOfferId: null,
    changedBookmarkOffer: {},
    relatedOffers: [],
    bookmarkOffers: [],
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    authInfo: {},
  },
};
