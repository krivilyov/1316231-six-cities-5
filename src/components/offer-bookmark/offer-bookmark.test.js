import React from "react";
import renderer from "react-test-renderer";
import {OfferBookmark} from "./offer-bookmark";
import configureMockStore from "redux-mock-store";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import {testInitialState} from "../../test-data";

const mockStore = configureMockStore();
const store = mockStore(testInitialState);

const noop = () => {};
describe(`OfferBookmark should render correctly`, () => {
  it(`OfferBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={false}`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferBookmark
                offerId={`1`}
                isAuthorizedStatus={false}
                offerBookmarkStatus={false}
                onChangeBookmark={noop}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={true}`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferBookmark
                offerId={`1`}
                isAuthorizedStatus={true}
                offerBookmarkStatus={true}
                onChangeBookmark={noop}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={true}`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferBookmark
                offerId={`1`}
                isAuthorizedStatus={false}
                offerBookmarkStatus={true}
                onChangeBookmark={noop}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={false}`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferBookmark
                offerId={`1`}
                isAuthorizedStatus={true}
                offerBookmarkStatus={false}
                onChangeBookmark={noop}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
