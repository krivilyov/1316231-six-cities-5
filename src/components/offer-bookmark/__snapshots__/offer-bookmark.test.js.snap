// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferBookmark should render correctly OfferBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={false} 1`] = `
<a
  className="property__bookmark-button button"
  href="/login"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height="33"
    width="31"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</a>
`;

exports[`OfferBookmark should render correctly OfferBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={true} 1`] = `
<a
  className="property__bookmark-button  property__bookmark-button--active button"
  href="/login"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height="33"
    width="31"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</a>
`;

exports[`OfferBookmark should render correctly OfferBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={false} 1`] = `
<button
  className="property__bookmark-button button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height="33"
    width="31"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</button>
`;

exports[`OfferBookmark should render correctly OfferBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={true} 1`] = `
<button
  className="property__bookmark-button  property__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height="33"
    width="31"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</button>
`;
